extension-common/pxfconfigparser.c

```diff
@@ -0,0 +1,661 @@
+                                   "at line %d of file '%s'",
+                                   ctx->key, ctx->value, ctx->line_number, ctx->filepath),
+                            errhint("please make sure the value is valid port numbers separated by ','")));
+        }
+
+        ListCell *lc;
+        foreach (lc, ports)
+        {
+            int port = parse_int(lfirst(lc), ctx);
+            group->ports = lappend_int(group->ports, port);
+        }
+    }
+    else if (strcmp(property_name, "ssl") == 0)
+    {
+        group->is_ssl = parse_bool(value_copy, ctx);
+    }
+    else
+    {
+        ereport(DEBUG1,
+                (errmsg("ignoring unknown configuration property '%s' in config file '%s' at line %d", ctx->key, ctx->filepath, ctx->line_number)));
+    }
+
+    pfree(value_copy);
+}
+
+/*
+ * Add a new pxf_cluster entry with a name of cluster_name into the deployment
+ * Expect deployment to be a valid pointer to pxf_deployment
+ * Expect cluster_name to be the same as the cluster's folder name
+ */
+static pxf_cluster *
+create_pxf_cluster(pxf_deployment *deployment, const char *cluster_name)
+{
+    pxf_cluster *cluster = palloc0(sizeof(pxf_cluster));
+    cluster->name = pstrdup(cluster_name);
+
+    deployment->clusters = lappend(deployment->clusters, cluster);
+    return cluster;
+}
+
+/*
+ * Clean up all allocated memory in pxf_deployment struct instance
+ * This function should be called when the pxf_deployment struct is no longer needed
+ */
+void clean_pxf_deployment(pxf_deployment *deployment)
+{
+    if (deployment == NULL)
+    {
+        return;
+    }
+    if (deployment->clusters != NULL)
+    {
+        ListCell *lc;
+        foreach (lc, deployment->clusters)
+        {
+            pxf_cluster *cluster = lfirst(lc);
+            clean_pxf_cluster(cluster);
+            pfree(cluster);
+        }
+        list_free(deployment->clusters);
+        deployment->clusters = NULL;
+    }
+    pfree(deployment);
+}
+
+/*
+ * Helper function called in deployment_cleanup to clean up a pxf_cluster struct
+ */
+static void
+clean_pxf_cluster(pxf_cluster *cluster)
+{
+    if (cluster == NULL)
+    {
+        return;
+    }
+
+    if (cluster->name)
+    {
+        pfree(cluster->name);
+    }
+
+    if (cluster->endpoint)
+    {
+        pfree(cluster->endpoint);
+    }
+
+    if (cluster->groups != NULL)
+    {
+        ListCell *lc;
+
+        foreach (lc, cluster->groups)
+        {
+            pxf_service_group *group = lfirst(lc);
+            // Clean up each service group
+            if (group->ports != NULL)
+            {
+                list_free(group->ports);
+                group->ports = NULL;
+            }
+            pfree(group->name);
+        }
+        list_free_deep(cluster->groups);
+        cluster->groups = NULL;
+    }
+
+    if (cluster->hosts != NULL)
+    {
+        ListCell *lc;
+        foreach (lc, cluster->hosts)
+        {
+            pxf_host *host = lfirst(lc);
```

Comment:
```suggestion
            char *host = lfirst(lc);
```
why not make it `char*` since there are no other attributes for the host ? Then you can just do `list_clean_deep` here instead of iterating over hosts list.

Reply:
> why not make it `char*` since there are no other attributes for the host ? Then you can just do `list_clean_deep` here instead of iterating over hosts list.

I followed the same data struct in Golang. Initially in Golang code I made it a string but the team gave me a feedback to make it a separate struct for consistency and extendability.